src/dash-table/components/ControlledTable/index.tsx

```diff
@@ -363,69 +364,88 @@ export default class ControlledTable extends PureComponent<ControlledTableProps>
-            // we are selecting left, move the right side closer to active_cell
-            removeCells = selectedRows.map(row => [row, maxCol]);
-        } else if (selectingLeft && minCol > 0) {
-            // Otherwise increase the selection left if possible
-            targetCells = selectedRows.map(row => [row, minCol - 1]);
-        } else if (selectingRight && (active_cell as any)[1] > minCol) {
-            // If there are selections to the left of the active cell and
-            // we are selecting right, move the left side closer to active_cell
-            removeCells = selectedRows.map(row => [row, minCol]);
-        } else if (selectingRight && maxCol + 1 <= columns.length - 1) {
-            // Otherwise move selection right if possible
-            targetCells = selectedRows.map(row => [row, maxCol + 1]);
+        let startRow = start_cell && start_cell.row;
+        let startCol = start_cell && start_cell.column;
+        let endRow = end_cell && end_cell.row;
+        let endCol = end_cell && end_cell.column;
+
+        if (selectingDown) {
+            if (active_cell.row > minRow) {
+                minRow++;
+                endRow = minRow;
+            } else if (maxRow < viewport.data.length - 1) {
+                maxRow++;
+                endRow = maxRow;
+            }
+        } else if (selectingUp) {
+            if (active_cell.row < maxRow) {
+                maxRow--;
+                endRow = maxRow;
+            } else if (minRow > 0) {
+                minRow--;
+                endRow = minRow;
+            }
+        } else if (selectingRight) {
+            if (active_cell.column > minCol) {
+                minCol++;
+                endCol = minCol;
+            } else if (maxCol < columns.length - 1) {
+                maxCol++;
+                endCol = maxCol;
+            }
+        } else if (selectingLeft) {
+            if (active_cell.column < maxCol) {
+                maxCol--;
+                endCol = maxCol;
+            } else if (minCol > 0) {
+                minCol--;
+                endCol = minCol;
+            }
+        } else {
+            return;
         }
 
-        const newSelectedCell = R.without(
-            removeCells,
-            R.uniq(R.concat(targetCells, selected_cells))
+        const finalSelected = makeSelection(
+            {minRow, maxRow, minCol, maxCol},
+            columns, viewport
         );
-        setProps({
+
+        const newProps: Partial<ICellFactoryProps> = {
             is_focused: false,
-            selected_cells: newSelectedCell
-        });
+            end_cell: makeCell(endRow, endCol, columns, viewport),
+            selected_cells: finalSelected
+        };
+
+        const newStartRow = endRow === minRow ? maxRow : minRow;
+        const newStartCol = endCol === minCol ? maxCol : minCol;
+
+        if (startRow !== newStartRow || startCol !== newStartCol) {
+            newProps.start_cell = makeCell(
```

Comment:
updating `start_cell` only for that strange situation of reshaping the selection after moving `active_cell`